text/0074-global-init.md

```diff
@@ -0,0 +1,176 @@
+each SDK component to be intialized separately when the process
+starts.  The methods shall be declared in a separate API package,
+e.g., named `global`:
+
+```golang
+package global
+
+import (
+    "go.opentelemetry.io/otel/api/context/propagation"
+    "go.opentelemetry.io/otel/api/metric"
+    "go.opentelemetry.io/otel/api/trace"
+)
+
+// SetMeterProvider initializes the global Meter provider.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Meter provider, the default global
+// Meter provider acts as a "forwarding" SDK.  The default global
+// Meter provider will begin forwarding to the installed Meter
+// provider once this is called.
+func SetMeterProvider(metric.Provider) { ... }
+
+// SetPropagators initializes the global Propagators instance.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Meter provider, the default global
+// Propagators instance performs pass-through W3C Traceparent and
+// Correlation-Context propagation.
+func SetPropagators(propagation.Propagators) { ... }
+
+// SetPropagators initializes the global Tracer provider.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Tracer provider, the default global
+// Tracer provider acts as a "forwarding" SDK.  The default global
+// Tracer provider will begin forwarding to the installed Tracer
+// provider once this is called.
+func SetTraceProvider(trace.Provider) { ... }
+```
+
+#### Requirements
+
+We anticipate third party libraries using the global instances before
+they are installed, and we wish for references obtained through these
+instances to become functional once the corresponding implementation
+is initialized.  The default instances returned by the global getters
+for Tracer provider, Meter provider, and Propagators must "forward" to
+the real SDK implementation once it is installed.
+
+#### Tracer
+
+Tracers obtained through the provider will become functional when the
+user's Tracer SDK is installed as the global instance.
+
+Spans started prior to installing the Tracer SDK will be No-op spans.
+Installing a Tracer SDK after starting a span via the default global
+instance does not change this behavior.
+
+#### Meter
+
+Meters obtained through the provider will become functional when the
+users's Meter SDK is installed as the global instance.
+
+Metric events will be dropped until the Meter SDK is installed.
+
+#### Propagators
+
+The default global Propagators instance will by default perform
+pass-through W3C Traceparent and Correlation-Context propagation.
+
+The default global Propagators instance will begin forwarding to the
+user's Propagators when it is installed as the global instance.
+
+## Trade-offs and mitigations
+
+### Testing support
+
+Testing should be performed without depending on the global SDK, if
```

Comment:
Yes! Hopefully tests can depend on specific implementations of the specific pieces of the SDK, rather than having everything require the global instance, and a complex global-reset dance.

Reply:
In the prototype Go PR, I made an _internal_ method for testing the global variables, but it can only be used from its own package sub-tree, so it's not generally available.

Reply:
👍

Reply:
Playing devil's advocate, what is an example where testing can explicitly inject the SDK but the actual instrumentation cannot do that?

Reply:
I don't think there are any, but if people build instrumentation always just via accessing the globals, then unit testing gets difficult. I guess they're making their own troubles, though!

Reply:
Agreed. Testing against the global SDK should be kept at minimum.

Reply:
I'm not sure this answer's Yuri's question, but I find it difficult to imagine a module of code that both (a) wants to test their instrumentation and (b) wants to use the global SDK. It will be difficult to test instrumentation against the global SDK because you'll have to install an SDK such as a mock. But the requirements here say you can only install an SDK once. You can still write tests in this way, but they have to be single-threaded.